### PR TITLE
[CELEBORN-720][SPARK] Correct metric peakExecutionMemory of SortBasedShuffleWriter

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
Currently SortBasedShuffleWriter won't update peakMemoryUsedBytes, this pr support this.

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
ping @pan3793

**Comment by pan3793** [comment]
cc @cfmcgrady too

File: client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SortBasedShuffleWriter.java [hunk 1]
```diff
@@ -377,7 +405,7 @@ private void updateMapStatus() {
   @Override
   public Option<MapStatus> stop(boolean success) {
     try {
-      taskContext.taskMetrics().incPeakExecutionMemory(peakMemoryUsedBytes);
+      taskContext.taskMetrics().incPeakExecutionMemory(getPeakMemoryUsedBytes());
```
Comment (pan3793) [inline client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SortBasedShuffleWriter.java:408]: The variable `private long peakMemoryUsedBytes = 0;` defined in this class could be deleted then?

Comment (AngersZhuuuu) [reply client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SortBasedShuffleWriter.java:408]: > The variable `private long peakMemoryUsedBytes = 0;` defined in this class could be deleted then?

No, still need

File: client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SortBasedShuffleWriter.java [hunk 2]
```diff
@@ -189,6 +189,34 @@ public SortBasedShuffleWriter(
...
+    if (pipelined) {
+      for (int i = 0; i < pushers.length; i++) {
+
+        if (pushers[i] != null) {
+          long mem = pushers[i].getPeakMemoryUsedBytes();
+          if (mem > peakMemoryUsedBytes) {
+            peakMemoryUsedBytes = mem;
+          }
+        }
+      }
```
Comment (pan3793) [inline client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SortBasedShuffleWriter.java:203]:
```suggestion
      for (SortBasedPusher pusher : pushers) {
        if (pusher != null) {
          long mem = pusher.getPeakMemoryUsedBytes();
          if (mem > peakMemoryUsedBytes) {
            peakMemoryUsedBytes = mem;
          }
        }
      }
```

**Review by pan3793 — APPROVED** [review]
LGTM, even some variables are not counted

```
    serBuffer = new OpenByteArrayOutputStream(DEFAULT_INITIAL_SER_BUFFER_SIZE);
    serOutputStream = serializer.serializeStream(serBuffer);

    this.mapStatusLengths = new LongAdder[numPartitions];
    for (int i = 0; i < numPartitions; i++) {
      this.mapStatusLengths[i] = new LongAdder();
    }
    tmpRecords = new long[numPartitions];
```

File: client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SortBasedShuffleWriter.java (hunk 1 above)
Comment (pan3793) [reply client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SortBasedShuffleWriter.java:406]: I see

**Review by cfmcgrady — APPROVED** [review]
LGTM

**Comment by pan3793** [comment]
Thanks, merged to main/0.3

**Pull request closed without merging**